### PR TITLE
gh-115986 Improve pprint docs formatting

### PR DESCRIPTION
**Pull request by Privat33r-dev**
The change improves readability.
Suggested in the #116085 PR discussion.

* Issue: gh-115986

----
📚 Documentation preview 📚: https://cpython-previews--117401.org.readthedocs.build/en/117401/library/pprint.html

### CODE REVIEW

**Comment by erlend-aasland** [comment]
If we're putting the params list in the `pprint.pprint` function docs, we could just as well use the standardised Sphinx param list markup (`:param:`, `:type:`, etc.). We'll probably refer to `pprint.pprint` from the `pprint.pp` and the `PrettyPrinter` docs anyway, so using the standard markup seems like a better way to me. cc. @alexwaygood @hugovk

**Comment by Privat33r-dev** [comment]
@erlend-aasland, I do completely agree that `:param:` will look better, especially considering the fact that we want to keep parameters' types as well and `:param:` look much more suitable for this purpose. I was hesitant to express my opinion earlier as everyone seemed to agree on table's solution, but now I see that it made more sense to speak my mind.

Moreover, I want to address the whole "pp vs pprint" thing. If we are writing docs from the developer's perspective, it would seem more logical to write the full description in one function's docs and just mention in another that it's an alias (please, ignore formatting and minor details):

```md
pprint.pp(object, stream=None, indent=1, width=80, depth=None, *, compact=False, sort_dicts=False, underscore_numbers=False)
Prints the formatted representation of object on stream, followed by a newline. If stream is None, [sys.stdout] is used. This may be used in the interactive interpreter instead of the [print()] function for inspecting values (you can even reassign print = pprint.pp for use within a scope).

The configuration parameters stream, indent, width, depth, compact, sort_dicts and underscore_numbers are passed to the [PrettyPrinter] constructor and their meanings are as described in its documentation below.

** PrettyPrinter PARAMS HERE **

pprint.pprint(object, stream=None, indent=1, width=80, depth=None, *, compact=False, sort_dicts=True, underscore_numbers=False)
Alias to pprint.pp. Note that sort_dicts is True by default and you might want to use [pp()] instead where it is False by default.
```

It feels like it would make the most sense, because here we would give all the usage info on pprint.pp and only then mention "pprint.pprint" and other stuff.

**Comment by erlend-aasland** [comment]
@Privat33r-dev, do you have a draft PR of your suggestion?

**Comment by Privat33r-dev** [comment]
> @Privat33r-dev, do you have a draft PR of your suggestion?

not yet, but I can make (or rather edit existing) one if my proposed change sounds good

**Comment by erlend-aasland** [comment]
> not yet, but I can make (or rather edit existing) one if my proposed change sounds good

I think a draft PR would be interesting, but only if you have time available for it; I find it easier to decide on doc layouts when they are rendered :)

**Comment by Privat33r-dev** [comment]
> > not yet, but I can make (or rather edit existing) one if my proposed change sounds good
> 
> I think a draft PR would be interesting, but only if you have time available for it; I find it easier to decide on doc layouts when they are rendered :)

done, you can check it out :)

**Comment by erlend-aasland** [comment]
Thanks! I like this very much. We _could_ use Sphinx placeholders to minimise the duplication. I'm not sure if it is worth it, though, because you'd have to create one placeholder for each parameter description.

**Comment by Privat33r-dev** [comment]
There are 2 differences between the params table:
1. "object" presence as the first `param`
2. (the default) placement for `sort_dicts` param

Even if we technically can make a placeholder for table (by "table" I mean params list), the first difference is resolved by putting `:param:` before the table, but the second difference could be resolved only by moving the table to `pprint.pprint` or by somehow replacing the content of the placeholder (which is kind of hacky).

We can keep it DAMP for now since there is low change of significant changes in the `pprint.pprint` and in case if there are some changes (which will unlikely involve parameters), we can add comment to rst that edit should be made twice.

**Comment by Privat33r-dev** [comment]
Btw, my recent change:
**Move code out of parameter**
<details><summary>Screenshots</summary>
<p>
<h2>Before</h2>

<h2>After</h2>

</p>
</details>
<b>Move code out of "changed in version" and move the latter down</b>
<details><summary>Screenshots</summary>
<p>
<h2>Before</h2>

<h2>After</h2>

</p>
</details>

**Comment by erlend-aasland** [comment]
Moving the example code out of the parameter list sounds good. IMO we should never embed code examples in parameter lists, since it is (IMO) visually cluttering.

**Comment by Privat33r-dev** [comment]
> Moving the example code out of the parameter list sounds good. IMO we should never embed code examples in parameter lists, since it is (IMO) visually cluttering.

I am not sure how to fix the tests though. Can you give an advise? This change will likely do the trick:
image

But I'd prefer to keep the indentation and avoid "usage example" and similar things to keep doc consistency.

**Comment by erlend-aasland** [comment]
> I am not sure how to fix the tests though

You need to adjust the indentation so it matches the expected output; AFAIK, there is no other way.

**Comment by Privat33r-dev** [comment]
> > I am not sure how to fix the tests though
> 
> You need to adjust the indentation so it matches the expected output; AFAIK, there is no other way.

putting additional newline between "usage example" and code fixed the indentation. I tried to use comment (`.. example\n\n{code}`), but it made example visually disappear. I guess we go with "usage example"s now. But, I guess, it makes the doc more `explicit` now :)

**Comment by Privat33r-dev** [comment]
I figured out that the problem wasn't in the indentation of the block code but because my IDE decided to align code inside of the code blocks on _\<shift\>_+_\<tab\>_.

**Comment by Privat33r-dev** [comment]
image
Well, after watching the Steven He's "Failure Management" [course](https://www.youtube.com/watch?v=TsSP-nHjZ0Q), I can confidently say that the problem is still in indentation and reducing amount of failures from 3 to 1 is a partial success.

**Comment by Privat33r-dev** [comment]
Failure has been successfully managed. Ready for review :)

P.S. Why does `doctest` job take 16 minutes? Diving a bit deeper I noticed that `doctest.py` does not parallelize jobs, it might be the problem.

File: Doc/library/pprint.rst [hunk 1]
```diff
@@ -36,52 +36,85 @@ Dictionaries are sorted by key before the display is computed.
...
-   :func:`~pprint.pp` instead where it is ``False`` by default.
 
 .. function:: pformat(object, indent=1, width=80, depth=None, *, \
                       compact=False, sort_dicts=True, underscore_numbers=False)
 
    Return the formatted representation of *object* as a string.  *indent*,
    *width*, *depth*, *compact*, *sort_dicts* and *underscore_numbers* are
    passed to the :class:`PrettyPrinter` constructor as formatting parameters
-   and their meanings are as described in its documentation below.
+   and their meanings are as described in the documentation above.
```
Comment (encukou) [inline Doc/library/pprint.rst:116]: The documentation for `PrettyPrinter` is still below. Maybe this should be:
```
The arguments have the same meaning as for :func:`~pprint.pprint`.
Note that *sort_dicts* defaults to ``True``.
```

File: Doc/library/pprint.rst [hunk 2]
```diff
@@ -36,52 +36,85 @@ Dictionaries are sorted by key before the display is computed.
...
+      the next contained level is replaced by ``...``.
+      If ``None`` (the default), there is no constraint
+      on the depth of the objects being formatted.
+   :type depth: int | None
+
+   :param stream:
+      A file-like object to which the output will be written
+      by calling its :meth:`!write` method.
+      If ``None`` (the default), :data:`sys.stdout` is used.
+   :type stream: :term:`file-like object` | None
```
Comment (encukou) [inline Doc/library/pprint.rst]: This doesn't match the order of the positional arguments. Could you move it just after `object`?

File: Doc/library/pprint.rst [hunk 3]
```diff
@@ -120,51 +153,74 @@ Functions
...
-   *stream* (default :data:`!sys.stdout`) is a :term:`file-like object` to
-   which the output will be written by calling its :meth:`!write` method.
-   If both *stream* and :data:`!sys.stdout` are ``None``, then
-   :meth:`~PrettyPrinter.pprint` silently returns.
-
-   Other values configure the manner in which nesting of complex data
-   structures is displayed.
+   Construct a :class:`PrettyPrinter` instance.
+
+   :param int indent:
```
Comment (encukou) [inline Doc/library/pprint.rst]: Does it make sense to repeat everything? I don't think there'd be any harm in directing the user to the prominent block at the beginning of the docs.
```
Arguments have the same meaning as for :func:`~pprint.pp`.
Note that they are in a different order, and that *sort_dicts* defaults to ``True``.
```

File: Doc/library/pprint.rst [hunk 4]
```diff
@@ -120,51 +153,74 @@ Functions
...
+      will be formatted on each output line.
+
+   :param bool sort_dicts:
+      If ``True`` (the default), dictionaries will be formatted with
+      their keys sorted, otherwise they will be displayed in insertion order.
+
+   :param bool underscore_numbers:
+      If ``True``,
+      integers will be formatted with the ``_`` character for a thousands separator,
+      otherwise underscores are not displayed (the default).
```
Comment (Privat33r-dev) [inline Doc/library/pprint.rst]:
```suggestion
   Arguments have the same meaning as for :func:`~pprint.pp`.
   Note that they are in a different order, and that *sort_dicts* defaults to ``True``.
```

File: Doc/library/pprint.rst (hunk 1 above)
Comment (Privat33r-dev) [reply Doc/library/pprint.rst:117]: Thank you for your suggestion. Addressed in aeb9488674dba0281fabfed5718958620aad1ab6

File: Doc/library/pprint.rst (hunk 2 above)
Comment (Privat33r-dev) [inline Doc/library/pprint.rst]:
```suggestion
   :param object:
      The object to be printed.

   :param stream:
      A file-like object to which the output will be written
      by calling its :meth:`!write` method.
      If ``None`` (the default), :data:`sys.stdout` is used.
   :type stream: :term:`file-like object` | None

   :param int indent:
      The amount of indentation added for each nesting level.

   :param int width:
      The desired maximum number of characters per line in the output.
      If a structure cannot be formatted within the width constraint,
      a best effort will be made.

   :param depth:
      The number of nesting levels which may be printed.
      If the data structure being printed is too deep,
      the next contained level is replaced by ``...``.
      If ``None`` (the default), there is no constraint
      on the depth of the objects being formatted.
   :type depth: int | None
```

File: Doc/library/pprint.rst (hunk 3 above)
Comment (Privat33r-dev) [reply Doc/library/pprint.rst]: Thank you for your suggestion. Addressed in https://github.com/python/cpython/commit/aeb9488674dba0281fabfed5718958620aad1ab6

File: Doc/library/pprint.rst (hunk 1 above)
Comment (Privat33r-dev) [reply Doc/library/pprint.rst:116]: That's why "its" was changed to "the", as arguments description is above (in the `pp.pprint` description).

File: Doc/library/pprint.rst (hunk 2 above)
Comment (Privat33r-dev) [reply Doc/library/pprint.rst]: Good catch, Addressed in 4283ad1c019c5b41f9d3f100a029b37eeb0c01f5

**Comment by hugovk** [comment]
I've merged `main` in, should fix the CI failure (`check-warnings.py: error: unrecognized arguments: --fail-if-new-news-nit`).

**Comment by Privat33r-dev** [comment]
> I've merged `main` in, should fix the CI failure (`check-warnings.py: error: unrecognized arguments: --fail-if-new-news-nit`).

Thanks. It worked :)

**Review by encukou — APPROVED** [review]

**Comment by encukou** [comment]
Thank you!

**Comment by erlend-aasland** [comment]
Next time, remember to edit the commit message upon merging, @encukou :)

Thanks for the contribution, Kerim!

**Pull request merged**